### PR TITLE
Porting adapter from lit-llama

### PR DESCRIPTION
**Pull request by ArturK-85**
Fine tuning adapter on consumer grade graphic card ~12GB VRam 

Part of #20 
The files were copied and adapted from the Lit-LLaMA repo: https://github.com/Lightning-AI/lit-llama

### CODE REVIEW

**Comment by awaelchli** [comment]
Hey @ArturK-85 thanks for working on this! I'm currently reviewing and running this script.

File: finetune_adapter.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+                          #"reduce_scatter":True,
+                          #"contiguous_gradients": True,
+                          #"round_robin_gradients": True,
+                          "offload_optimizer": {
+                             "device": "cpu",
+                             "pin_memory":True,
+                          },
+                          "offload_param": {
+                             "device": "cpu",
+                          }
```
Comment (awaelchli) [inline finetune_adapter.py:53]: I propose that we keep this at just the base settings with stage 2. The rest could be tuned in a follow-up, i.e., the cpu offload would require switching the optimizer.

**Comment by awaelchli** [comment]
The finetuning script works. Here are some sample instructions and responses from the model:

> What food do lamas eat?

Lamas are vegetarians who supplement their diet with animal protein and/or fish. They also frequently consume wild seafood and plant-based protein sources. Lamas also consume a wide range of fruits and vegetables, and occasionally consume some dairy products. Lamas also typically avoid meat, seafood, and eggs because they are high in saturated fat and cholesterol. Lamas are also helped by medical advice and guidance from their spiritual leader, who can guide them to take care of themselves.


> Recommend a book to read on the weekend. Explain the reason. I like fiction and horror.

I prefer reading romance stories over thrillers and thrillers over thrillers. One of my favorite romance stories is "The Long Way Round" by Anthony Robins and is a perfect example of how the author conveys both the romance and the suspense of the story.

> Write a Python program that tests whether a number is prime.

```
def check_prime(n):
    if n != 1:
        return False
    for i in range(2,n):
        if n % i == 0:
            return False
    return True
```

This either needs to train longer or we need to tweak the hyperparameters a bit more. I'll investigate but functionally I think the PR is ready. Great work!

File: lit_stablelm/tokenizer.py [hunk 2]
```diff
@@ -25,8 +25,26 @@ def token_to_id(self, token: str) -> int:
             raise ValueError(f"token {token!r} not found in the collection.")
         return id_
 
-    def encode(self, string: str, device: Optional[torch.device] = None) -> torch.Tensor:
+    def encode(
+        self, 
+        string: str, 
+        device: Optional[torch.device] = None, 
+        bos: bool = False,
```
Comment (awaelchli) [inline lit_stablelm/tokenizer.py:32]: Note: I set bos=False as default here to make the existing tokenizer test pass in this repo. This is different from the Lit-LLaMA version. 

@carmocca @lantiga Let me know if this is undesired, we can also adapt the test.

File: finetune_adapter.py (hunk 1 above)
Comment (lantiga) [reply finetune_adapter.py:53]: I agree with this

File: lit_stablelm/tokenizer.py (hunk 2 above)
Comment (lantiga) [reply lit_stablelm/tokenizer.py:32]: This is fine, I believe GPT only has the padding token

Comment (carmocca) [reply lit_stablelm/tokenizer.py:32]: Do we need the token arguments at all? Isn't this managed by the HFTokenizer internally?

BTW the changes here should be reflected in `test_tokenizer.py`

File: finetune_adapter.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        "offload_optimizer": {"device": "cpu", "pin_memory": True},
+        "offload_param": {"device": "cpu"},
+    },
+}
+
+
+def main(
+    data_dir: str = "data/alpaca",
+    pretrained_path: str = "checkpoints/stabilityai/stablelm-base-alpha-3b/lit_model.pth",
+    out_dir: str = "out/adapter/alpaca",
```
Comment (carmocca) [inline finetune_adapter.py:57]: In this repo, we are using a single checkpoint dir for simplicity in passing arguments

```suggestion
    checkpoint_dir: Path = Path("checkpoints/stabilityai/stablelm-base-alpha-3b"),
```

**Review by ArturK-85 — COMMENTED** [review]
Single checkpoint_dir looks great!

**Review by ArturK-85 — COMMENTED** [review]
Looks great

**Review by carmocca — APPROVED** [review]

**Comment by lantiga** [comment]
Thank you @ArturK-85 !

**Comment by ArturK-85** [comment]
@lantiga 

Also thank you, I'm glad that the whole code works and new possibilities for edge devices have been unlocked

**Comment by ArturK-85** [comment]
@awaelchli @carmocca @lantiga

Thank you guys, great job!

File: finetune_adapter.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+def main(
+    data_dir: Path = Path("data/alpaca"), checkpoint_dir: Path = Path("checkpoints/stabilityai/stablelm-base-alpha-3b")
+):
+    check_valid_checkpoint_dir(checkpoint_dir)
+
+    fabric = L.Fabric(
+        accelerator="cuda",
+        devices=devices,
+        strategy=(DeepSpeedStrategy(config=ds_config) if devices > 1 else "auto"),
+        precision="bf16-mixed",
```
Comment (carmocca) [inline finetune_adapter.py:60]: Lit-llama uses `bf16-true` here after https://github.com/Lightning-AI/lit-llama/pull/201, which landed before this PR if I'm not mistaken.

@ArturK-85 did you set `mixed` here on purpose? If not, we should standardize
